File: examples/example-expo-53/modules/native-crash/android/src/main/java/expo/modules/nativecrash/NativeCrashModule.kt
package expo.modules.nativecrash

import android.os.Handler
import android.os.Looper
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class NativeCrashModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("NativeCrash")

        Function("crashNative") {
            // Re-throw on the main looper so the exception escapes Expo's call
            // handler (which would otherwise convert it into a JS error) and
            // reaches the JVM's uncaught-exception handler that posthog-android
            // hooks for native crash capture.
            Handler(Looper.getMainLooper()).post {
                crashOuterLayer()
            }
        }
    }
}

// Nested calls so the captured native stack trace has meaningful frames.
private fun crashOuterLayer() {
    crashMiddleLayer()
}

private fun crashMiddleLayer() {
    crashInnerLayer()
}

private fun crashInnerLayer() {
    throw RuntimeException("PostHog native Android test crash")
}

File: examples/example-expo-57/modules/native-crash/ios/NativeCrashModule.swift
import ExpoModulesCore

public class NativeCrashModule: Module {
    public func definition() -> ModuleDefinition {
        Name("NativeCrash")

        // `fatalError` is a Swift trap — it cannot be caught by Expo's call
        // handler, so the process terminates and posthog-ios' crash reporter
        // captures it (delivered on next launch).
        Function("crashNative") {
            CrashOuterLayer.process()
        }
    }
}

// Nested layers so the captured native stack trace has meaningful frames.
private enum CrashOuterLayer {
    static func process() {
        CrashMiddleLayer.handle()
    }
}

private enum CrashMiddleLayer {
    static func handle() {
        CrashInnerLayer.execute()
    }
}

private enum CrashInnerLayer {
    static func execute() {
        fatalError("PostHog native iOS test crash")
    }
}
